libraries/AP_AIS/AP_AIS.h

```diff
@@ -0,0 +1,139 @@
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#pragma once
+
+#include <AP_Param/AP_Param.h>
+#include <AP_SerialManager/AP_SerialManager.h>
+//#include <AP_HAL/AP_HAL.h>
+#include <AP_Common/AP_ExpandingArray.h>
+
+#ifndef HAL_AIS_ENABLED
+#define HAL_AIS_ENABLED !HAL_MINIMIZE_FEATURES
+#endif
+
+#if HAL_AIS_ENABLED
+
+#define AIVDM_BUFFER_SIZE 10
+
+class AP_AIS
+{
+public:
+    AP_AIS();
+
+    /* Do not allow copies */
+    AP_AIS(const AP_AIS &other) = delete;
+    AP_AIS &operator=(const AP_AIS&) = delete;
+
+
+    // return true if AIS is enabled
+    bool enabled() const;
+
+    // Initialize the AIS object and prepare it for use
+    void init();
+
+    // update AIS, expected to be called at 20hz
+    void update();
+
+    // send mavlink AIS message
+    void send(mavlink_channel_t chan);
+
+    // parameter block
+    static const struct AP_Param::GroupInfo var_info[];
+
+private:
+
+    // parameters
+    AP_Int8 _type;             // type of AIS recever
+    AP_Int16 _max_list;        // maximum number of vessels to track at once
+    AP_Int16 _time_out;        // time in seconds that a vessel will be dropped from the list
+    AP_Int16 _options;         // options bitmask
+
+    enum AISType {
+        AIS_NONE   = 0,
+        AIS_NMEA   = 1,
+    };
+
+    enum options {
+        AIS_OPTIONS_LOG_ALL_RAW         = 1<<0,
+        AIS_OPTIONS_LOG_UNSUPPORTED_RAW = 1<<1,
+        AIS_OPTIONS_LOG_DECODED         = 1<<2,
+    };
+
+    struct AIVDM {
+        uint8_t num;
+        uint8_t total;
+        uint8_t ID;
+        char payload[65];
+    };
+    AIVDM _incoming;
+    AIVDM _AIVDM_buffer[AIVDM_BUFFER_SIZE];
+
+    struct ais_vehicle_t {
+        mavlink_ais_vessel_t info;
+        uint32_t last_update_ms; // last time this was refreshed, allows timeouts
+        uint32_t last_send_ms; // last time this message was sent via mavlink, stops us spamming the link
+    };
+
+    // list of the vessels that are being tracked
+    AP_ExpandingArray<ais_vehicle_t> _list;
+
+    AP_HAL::UARTDriver *_uart;
+
+    uint16_t _send_index; // index of the last vessel send over mavlink
+
+    // removed the given index from the AIVDM buffer shift following elements
+    void buffer_shift(uint8_t i);
+
+    // find vessel in existing list, if not then return new index if possible
+    bool get_vessel_index(uint32_t mmsi, uint16_t &index, uint32_t lat = 0, uint32_t lon = 0);
+    void clear_list_item(uint16_t index);
+
+    // decode the payload
+    bool payload_decode(const char *payload);
+
+    // decode specific message types
+    bool decode_position_report(const char *payload, uint8_t type);
```

Comment:
@peterbarker might say we should have a WARN_IF_UNUSED on these methods returning bools.

Reply:
They're pretty cheap.